File: frontend/ios/App/CapApp-SPM/Sources/CapApp-SPM/CompanionBridgePlugin.swift
import Capacitor
import Foundation
import Security
import WebKit

public enum CompanionBridgeStorage {
    public static let accessTokenKey = "companion.accessToken"
    public static let refreshTokenKey = "companion.refreshToken"
    public static let apiBaseUrlKey = "companion.apiBaseUrl"
    public static let preferredLanguageKey = "companion.preferredLanguage"
    public static let didUpdateNotification = Notification.Name("CompanionBridgeDidUpdate")

    private static let accessService = "com.jefe.cacaradar.auth.access"
    private static let refreshService = "com.jefe.cacaradar.auth.refresh"
    private static let keychainAccount = "primary"

    public static func readAccessToken() -> String? {
        readKeychainValue(service: accessService)
    }

    public static func readRefreshToken() -> String? {
        readKeychainValue(service: refreshService)
    }

    public static func readApiBaseUrl() -> String? {
        UserDefaults.standard.string(forKey: apiBaseUrlKey)
    }

    public static func readPreferredLanguage() -> String? {
        UserDefaults.standard.string(forKey: preferredLanguageKey)
    }

    public static func persistAuthState(
        accessToken: String,
        refreshToken: String?,
        apiBaseUrl: String,
        preserveStoredRefreshToken: Bool = false
    ) throws {
        try persistKeychainValue(accessToken, service: accessService)
        let nextRefreshToken = refreshToken?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !preserveStoredRefreshToken || !nextRefreshToken.isEmpty {
            try persistKeychainValue(nextRefreshToken, service: refreshService)
        }

        let defaults = UserDefaults.standard
        defaults.removeObject(forKey: accessTokenKey)
        defaults.removeObject(forKey: refreshTokenKey)
        if apiBaseUrl.isEmpty {
            defaults.removeObject(forKey: apiBaseUrlKey)
        } else {
            defaults.set(apiBaseUrl, forKey: apiBaseUrlKey)
        }
        defaults.synchronize()
    }

    public static func clearAuthState() throws {
        try clearKeychainValue(service: accessService)
        try clearKeychainValue(service: refreshService)
        UserDefaults.standard.removeObject(forKey: accessTokenKey)
        UserDefaults.standard.removeObject(forKey: refreshTokenKey)
        UserDefaults.standard.removeObject(forKey: apiBaseUrlKey)
        UserDefaults.standard.synchronize()
    }

    public static func currentAuthState() -> [String: String] {
        [
            "accessToken": readAccessToken() ?? "",
            "apiBaseUrl": readApiBaseUrl() ?? "",
        ]
    }

    private static func baseQuery(service: String) -> [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: keychainAccount,
        ]
    }

    private static func persistKeychainValue(_ value: String, service: String) throws {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty {
            try clearKeychainValue(service: service)
            return
        }

        let data = Data(trimmed.utf8)
        var attributes = baseQuery(service: service)
        let updateStatus = SecItemUpdate(
            attributes as CFDictionary,
            [
                kSecValueData as String: data,
                kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
            ] as CFDictionary
        )
        if updateStatus == errSecSuccess {
            return
        }
        if updateStatus != errSecItemNotFound {
            throw keychainError(status: updateStatus, service: service)
        }

        attributes[kSecValueData as String] = data
        attributes[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        let addStatus = SecItemAdd(attributes as CFDictionary, nil)
        if addStatus != errSecSuccess {
            throw keychainError(status: addStatus, service: service)
        }
    }

    private static func readKeychainValue(service: String) -> String? {
        var query = baseQuery(service: service)
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne

        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        guard status == errSecSuccess else {
            return nil
        }
        guard let data = item as? Data else {
            return nil
        }
        let token = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines)
        return token?.isEmpty == false ? token : nil
    }

    private static func clearKeychainValue(service: String) throws {
        let status = SecItemDelete(baseQuery(service: service) as CFDictionary)
        if status != errSecSuccess && status != errSecItemNotFound {
            throw keychainError(status: status, service: service)
        }
    }

    private static func keychainError(status: OSStatus, service: String) -> NSError {
        let message = SecCopyErrorMessageString(status, nil) as String? ?? "Keychain error"
        return NSError(
            domain: "CompanionBridgeStorage",
            code: Int(status),
            userInfo: [
                NSLocalizedDescriptionKey: "\(message) (\(service))",
            ]
        )
    }
}

@objc(CompanionBridgePlugin)
public class CompanionBridgePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CompanionBridgePlugin"
    public let jsName = "CompanionBridge"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "syncAuthState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAuthState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "bootstrapSessionFromCookies", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "refreshAccessToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "syncPreferences", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearAuthState", returnType: CAPPluginReturnPromise),
    ]

    @objc func syncAuthState(_ call: CAPPluginCall) {
        let accessToken = call.getString("accessToken") ?? ""
        let refreshToken = call.getString("refreshToken")
        let apiBaseUrl = call.getString("apiBaseUrl") ?? ""
        let preserveStoredRefreshToken = call.getBool("preserveStoredRefreshToken") ?? false

        do {
            try CompanionBridgeStorage.persistAuthState(
                accessToken: accessToken,
                refreshToken: refreshToken,
                apiBaseUrl: apiBaseUrl,
                preserveStoredRefreshToken: preserveStoredRefreshToken
            )
            NotificationCenter.default.post(name: CompanionBridgeStorage.didUpdateNotification, object: nil)
            call.resolve()
        } catch {
            call.reject("Failed to store auth state securely", nil, error)
        }
    }

    @objc func getAuthState(_ call: CAPPluginCall) {
        let authState = CompanionBridgeStorage.currentAuthState()
        var payload = JSObject()
        payload["accessToken"] = authState["accessToken"] ?? ""
        payload["apiBaseUrl"] = authState["apiBaseUrl"] ?? ""
        call.resolve(payload)
    }

    @objc func bootstrapSessionFromCookies(_ call: CAPPluginCall) {
        let apiBaseUrl = normalizedApiBaseUrl(call.getString("apiBaseUrl"))
        guard let apiUrl = URL(string: apiBaseUrl), let host = apiUrl.host else {
            call.reject("Invalid API base URL")
            return
        }

        let cookieStore = bridge?.webView?.configuration.websiteDataStore.httpCookieStore
            ?? WKWebsiteDataStore.default().httpCookieStore
        cookieStore.getAllCookies { cookies in
            let matchingCookies = cookies.filter { cookie in
                guard let cookieHost = cookie.domain.trimmingCharacters(in: .whitespacesAndNewlines).lowercased().split(separator: ":").first else {
                    return false
                }
                return host.lowercased().hasSuffix(cookieHost.hasPrefix(".") ? String(cookieHost.dropFirst()) : String(cookieHost))
            }
            let cookieAccessToken = matchingCookies.first(where: { $0.name == "access_token" })?.value ?? ""
            let cookieRefreshToken = matchingCookies.first(where: { $0.name == "refresh_token" })?.value ?? ""
            let accessToken = cookieAccessToken.isEmpty ? (CompanionBridgeStorage.readAccessToken() ?? "") : cookieAccessToken
            let refreshToken = cookieRefreshToken.isEmpty ? (CompanionBridgeStorage.readRefreshToken() ?? "") : cookieRefreshToken

            do {
                try CompanionBridgeStorage.persistAuthState(
                    accessToken: accessToken,
                    refreshToken: cookieRefreshToken.isEmpty ? nil : refreshToken,
                    apiBaseUrl: apiBaseUrl,
                    preserveStoredRefreshToken: cookieRefreshToken.isEmpty
                )
                NotificationCenter.default.post(name: CompanionBridgeStorage.didUpdateNotification, object: nil)
                var payload = JSObject()
                payload["accessToken"] = accessToken
                payload["synced"] = !refreshToken.isEmpty
                call.resolve(payload)
            } catch {
                call.reject("Failed to bootstrap session from cookies", nil, error)
            }
        }
    }

    @objc func refreshAccessToken(_ call: CAPPluginCall) {
        let apiBaseUrl = normalizedApiBaseUrl(call.getString("apiBaseUrl"))
        Task {
            do {
                let refreshedToken = try await refreshNativeAccessToken(apiBaseUrl: apiBaseUrl)
                NotificationCenter.default.post(name: CompanionBridgeStorage.didUpdateNotification, object: nil)
                var payload = JSObject()
                payload["accessToken"] = refreshedToken
                call.resolve(payload)
            } catch {
                call.reject("Failed to refresh native access token", nil, error)
            }
        }
    }

    @objc func syncPreferences(_ call: CAPPluginCall) {
        let preferredLanguage = call.getString("preferredLanguage") ?? ""

        let defaults = UserDefaults.standard
        defaults.set(preferredLanguage, forKey: CompanionBridgeStorage.preferredLanguageKey)
        defaults.synchronize()
        NotificationCenter.default.post(name: CompanionBridgeStorage.didUpdateNotification, object: nil)

        call.resolve()
    }

    @objc func clearAuthState(_ call: CAPPluginCall) {
        do {
            try CompanionBridgeStorage.clearAuthState()
            NotificationCenter.default.post(name: CompanionBridgeStorage.didUpdateNotification, object: nil)
            call.resolve()
        } catch {
            call.reject("Failed to clear auth state", nil, error)
        }
    }

    private func normalizedApiBaseUrl(_ rawValue: String?) -> String {
        let trimmed = (rawValue ?? CompanionBridgeStorage.readApiBaseUrl() ?? "https://cacaradar.es/api")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let withoutTrailingSlash = trimmed.replacingOccurrences(of: "/+$", with: "", options: .regularExpression)
        if withoutTrailingSlash.isEmpty {
            return "https://cacaradar.es/api"
        }
        if withoutTrailingSlash.hasSuffix("/api") {
            return withoutTrailingSlash
        }
        return withoutTrailingSlash + "/api"
    }

    private func refreshNativeAccessToken(apiBaseUrl: String) async throws -> String {
        guard let refreshToken = CompanionBridgeStorage.readRefreshToken(), !refreshToken.isEmpty else {
            throw NSError(domain: "CompanionBridgePlugin", code: 401, userInfo: [NSLocalizedDescriptionKey: "No refresh token available"])
        }
        guard let url = URL(string: apiBaseUrl + "/auth/refresh") else {
            throw NSError(domain: "CompanionBridgePlugin", code: 400, userInfo: [NSLocalizedDescriptionKey: "Invalid API base URL"])
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("1", forHTTPHeaderField: "X-Native-App")
        request.setValue("ios", forHTTPHeaderField: "X-Platform")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "refresh_token": refreshToken,
        ])

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw NSError(domain: "CompanionBridgePlugin", code: 401, userInfo: [NSLocalizedDescriptionKey: "Refresh request failed"])
        }
        guard
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
            let accessToken = json["access_token"] as? String,
            !accessToken.isEmpty
        else {
            throw NSError(domain: "CompanionBridgePlugin", code: 500, userInfo: [NSLocalizedDescriptionKey: "No access token returned"])
        }

        try CompanionBridgeStorage.persistAuthState(
            accessToken: accessToken,
            refreshToken: nil,
            apiBaseUrl: apiBaseUrl,
            preserveStoredRefreshToken: true
        )
        return accessToken
    }
}

File: frontend/android/app/src/main/java/com/jefe/cacaradar/CompanionBridgePlugin.java
package com.jefe.cacaradar;

import android.content.Context;
import android.content.SharedPreferences;
import android.webkit.CookieManager;

import androidx.security.crypto.EncryptedSharedPreferences;
import androidx.security.crypto.MasterKey;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;

@CapacitorPlugin(name = "CompanionBridge")
public class CompanionBridgePlugin extends Plugin {
    public static final String PREFS_NAME = "caca_companion_bridge";
    private static final String SECURE_PREFS_NAME = "caca_secure_auth";
    public static final String ACCESS_TOKEN_KEY = "access_token";
    public static final String REFRESH_TOKEN_KEY = "refresh_token";
    public static final String API_BASE_URL_KEY = "api_base_url";
    public static final String PREFERRED_LANGUAGE_KEY = "preferred_language";

    static final class AuthStateSnapshot {
        final String accessToken;
        final String refreshToken;
        final String apiBaseUrl;

        AuthStateSnapshot(String accessToken, String refreshToken, String apiBaseUrl) {
            this.accessToken = accessToken;
            this.refreshToken = refreshToken;
            this.apiBaseUrl = apiBaseUrl;
        }
    }

    private SharedPreferences prefs() {
        Context context = getContext();
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }

    static SharedPreferences securePrefs(Context context) throws Exception {
        MasterKey masterKey = new MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build();
        return EncryptedSharedPreferences.create(
            context,
            SECURE_PREFS_NAME,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        );
    }

    static void persistAuthState(Context context, String accessToken, String refreshToken, boolean preserveStoredRefreshToken) throws Exception {
        SharedPreferences.Editor secureEditor = securePrefs(context)
            .edit()
            .putString(ACCESS_TOKEN_KEY, accessToken == null ? "" : accessToken);
        if (!preserveStoredRefreshToken || (refreshToken != null && !refreshToken.trim().isEmpty())) {
            secureEditor.putString(REFRESH_TOKEN_KEY, refreshToken == null ? "" : refreshToken);
        }
        secureEditor.apply();
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .remove(ACCESS_TOKEN_KEY)
            .remove(REFRESH_TOKEN_KEY)
            .apply();
    }

    static AuthStateSnapshot readAuthState(Context context) throws Exception {
        SharedPreferences secure = securePrefs(context);
        SharedPreferences general = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        String configuredBaseUrl = general.getString(API_BASE_URL_KEY, "https://cacaradar.es/api");
        String apiBaseUrl = configuredBaseUrl == null || configuredBaseUrl.trim().isEmpty()
            ? "https://cacaradar.es/api"
            : configuredBaseUrl.replaceAll("/+$", "");
        return new AuthStateSnapshot(
            secure.getString(ACCESS_TOKEN_KEY, ""),
            secure.getString(REFRESH_TOKEN_KEY, ""),
            apiBaseUrl
        );
    }

    static void clearStoredAuthState(Context context) throws Exception {
        securePrefs(context)
            .edit()
            .remove(ACCESS_TOKEN_KEY)
            .remove(REFRESH_TOKEN_KEY)
            .apply();
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .remove(ACCESS_TOKEN_KEY)
            .remove(REFRESH_TOKEN_KEY)
            .remove(API_BASE_URL_KEY)
            .apply();
    }

    @PluginMethod
    public void syncAuthState(PluginCall call) {
        String accessToken = call.getString("accessToken", "");
        String refreshToken = call.getString("refreshToken");
        String apiBaseUrl = call.getString("apiBaseUrl", "");
        boolean preserveStoredRefreshToken = call.getBoolean("preserveStoredRefreshToken", false);

        try {
            persistAuthState(getContext(), accessToken, refreshToken, preserveStoredRefreshToken);
            prefs()
                .edit()
                .putString(API_BASE_URL_KEY, apiBaseUrl)
                .apply();
            call.resolve(new JSObject());
        } catch (Exception exception) {
            call.reject("Failed to store auth state securely", exception);
        }
    }

    @PluginMethod
    public void getAuthState(PluginCall call) {
        try {
            AuthStateSnapshot authState = readAuthState(getContext());
            JSObject result = new JSObject();
            result.put("accessToken", authState.accessToken);
            result.put("apiBaseUrl", authState.apiBaseUrl);
            call.resolve(result);
        } catch (Exception exception) {
            call.reject("Failed to read auth state", exception);
        }
    }

    @PluginMethod
    public void bootstrapSessionFromCookies(PluginCall call) {
        getBridge().execute(() -> {
            try {
                String apiBaseUrl = normalizeApiBaseUrl(call.getString("apiBaseUrl", ""));
                String webOrigin = apiBaseUrl.endsWith("/api")
                    ? apiBaseUrl.substring(0, apiBaseUrl.length() - 4)
                    : apiBaseUrl;
                String cookieHeader = CookieManager.getInstance().getCookie(webOrigin);
                String cookieAccessToken = readCookieValue(cookieHeader, ACCESS_TOKEN_KEY);
                String cookieRefreshToken = readCookieValue(cookieHeader, REFRESH_TOKEN_KEY);
                AuthStateSnapshot existingState = readAuthState(getContext());
                String accessToken = cookieAccessToken == null || cookieAccessToken.trim().isEmpty()
                    ? existingState.accessToken
                    : cookieAccessToken;
                String refreshToken = cookieRefreshToken == null || cookieRefreshToken.trim().isEmpty()
                    ? existingState.refreshToken
                    : cookieRefreshToken;

                persistAuthState(
                    getContext(),
                    accessToken,
                    cookieRefreshToken == null || cookieRefreshToken.trim().isEmpty() ? null : refreshToken,
                    cookieRefreshToken == null || cookieRefreshToken.trim().isEmpty()
                );
                prefs()
                    .edit()
                    .putString(API_BASE_URL_KEY, apiBaseUrl)
                    .apply();

                JSObject result = new JSObject();
                result.put("accessToken", accessToken == null ? "" : accessToken);
                result.put("synced", refreshToken != null && !refreshToken.trim().isEmpty());
                call.resolve(result);
            } catch (Exception exception) {
                call.reject("Failed to bootstrap session from cookies", exception);
            }
        });
    }

    @PluginMethod
    public void refreshAccessToken(PluginCall call) {
        getBridge().execute(() -> {
            try {
                AuthStateSnapshot authState = readAuthState(getContext());
                if (authState.refreshToken == null || authState.refreshToken.trim().isEmpty()) {
                    call.reject("No refresh token available");
                    return;
                }

                String apiBaseUrl = normalizeApiBaseUrl(call.getString("apiBaseUrl", authState.apiBaseUrl));
                URL url = new URL(apiBaseUrl + "/auth/refresh");
                HttpURLConnection connection = (HttpURLConnection) url.openConnection();
                connection.setRequestMethod("POST");
                connection.setRequestProperty("Content-Type", "application/json");
                connection.setRequestProperty("X-Native-App", "1");
                connection.setRequestProperty("X-Platform", "android");
                connection.setDoOutput(true);

                JSONObject body = new JSONObject();
                body.put("refresh_token", authState.refreshToken);
                try (OutputStream os = connection.getOutputStream()) {
                    os.write(body.toString().getBytes(StandardCharsets.UTF_8));
                }

                int status = connection.getResponseCode();
                if (status < 200 || status >= 300) {
                    call.reject("Refresh request failed");
                    return;
                }

                StringBuilder responseBuilder = new StringBuilder();
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        responseBuilder.append(line);
                    }
                }
                JSONObject json = new JSONObject(responseBuilder.toString());
                String accessToken = json.optString("access_token", "");
                if (accessToken == null || accessToken.trim().isEmpty()) {
                    call.reject("No access token returned");
                    return;
                }

                persistAuthState(getContext(), accessToken, null, true);
                prefs()
                    .edit()
                    .putString(API_BASE_URL_KEY, apiBaseUrl)
                    .apply();

                JSObject result = new JSObject();
                result.put("accessToken", accessToken);
                call.resolve(result);
            } catch (Exception exception) {
                call.reject("Failed to refresh native access token", exception);
            }
        });
    }

    @PluginMethod
    public void syncPreferences(PluginCall call) {
        String preferredLanguage = call.getString("preferredLanguage", "");

        prefs()
            .edit()
            .putString(PREFERRED_LANGUAGE_KEY, preferredLanguage)
            .apply();

        call.resolve(new JSObject());
    }

    @PluginMethod
    public void clearAuthState(PluginCall call) {
        try {
            clearStoredAuthState(getContext());
            call.resolve(new JSObject());
        } catch (Exception exception) {
            call.reject("Failed to clear auth state", exception);
        }
    }

    private static String normalizeApiBaseUrl(String rawValue) {
        String trimmed = rawValue == null ? "" : rawValue.trim();
        if (trimmed.isEmpty()) {
            return "https://cacaradar.es/api";
        }
        String normalized = trimmed.replaceAll("/+$", "");
        if (normalized.endsWith("/api")) {
            return normalized;
        }
        return normalized + "/api";
    }

    private static String readCookieValue(String cookieHeader, String cookieName) {
        if (cookieHeader == null || cookieHeader.trim().isEmpty()) {
            return "";
        }
        String[] cookies = cookieHeader.split(";");
        for (String cookie : cookies) {
            String[] parts = cookie.trim().split("=", 2);
            if (parts.length == 2 && cookieName.equals(parts[0].trim())) {
                return parts[1].trim();
            }
        }
        return "";
    }
}
